extruct/__init__.py

```diff
@@ -1,16 +1,39 @@
+import logging
+import functools
 from lxml.html import fromstring
-
 from extruct.jsonld import JsonLdExtractor
 from extruct.rdfa import RDFaExtractor
 from extruct.w3cmicrodata import MicrodataExtractor
+from extruct.opengraph import OpenGraphExtractor
+from extruct.microformat import MicroformatExtractor
 from extruct.xmldom import XmlDomHTMLParser
 
 
-def extract(htmlstring, url='http://www.example.com/', encoding="UTF-8"):
+logger = logging.getLogger(__name__)
+
+
+def extract(htmlstring, url='http://www.example.com/', encoding="UTF-8", 
+            syntaxes="all", schema_context='http://schema.org'):
     domparser = XmlDomHTMLParser(encoding=encoding)
     tree = fromstring(htmlstring, parser=domparser)
-    return {name: extractor.extract_items(tree, url=url)
-            for name, extractor in (
-                ('json-ld', JsonLdExtractor()),
-                ('microdata', MicrodataExtractor()),
-                ('rdfa', RDFaExtractor()))}
\ No newline at end of file
+    if syntaxes == 'all':
+        syntaxes = ['microdata', 'jsonld', 'opengraph', 'microformat', 'rdfa']
+    processors = []
+    if 'microdata' in syntaxes:
+        processors.append(('microdata', MicrodataExtractor().extract_items))
+    if 'jsonld' in syntaxes:
+        processors.append(('jsonld', JsonLdExtractor().extract_items))
+    if 'opengraph' in syntaxes:
+        processors.append(('opengraph', OpenGraphExtractor().extract_items))
+    if 'microformat' in syntaxes:
+        processors.append(('microformat', MicroformatExtractor().extract_items))
+    if 'rdfa' in syntaxes:
+        processors.append(('rdfa', RDFaExtractor().extract_items))           
+
+    output = {}
+    for label, extract in processors:
+        try:
+            output[label] = [obj for obj in extract(document=tree, url=url, html=html)]
+        except Exception:
+            logger.exception("Failed to parse %s", url)
```

Comment:
why are exceptions silenced?

Reply:
I missed this comment before. I copied this bit from dse-microdata and I thought it was useful to keep it because in production if there is an error with parsing one syntax for a specific domain you don't necessarily want extraction to crash on all pages. As Konstantin mentioned very often webpages are annotated with multiple syntaxes and information is duplicated so the fact that, say rdfa, crashes on a page does not mean we need to stop on it

Reply:
Yeah, I agree it makes sense to provide a way to ignore errors, but I think it'd be good to have an option to give users an access to the outcome (probably enabled by default). A function argument like `errors="log" / "ignore", "strict"`, with "strict" being a default value, can work; this would be similar to Python's "encode" function. 

Another option is to set `output[label]` to the exception object in case of error, but I like it less, as it looks like a trick.